src/common/cmd.c

```diff
@@ -493,3 +546,37 @@ static cmdstat_t cmd_wait_fn(i32 argc, const char** argv)
         return cmdstat_ok;
     }
 }
+
+static cmdstat_t cmd_cmds_fn(i32 argc, const char** argv)
+{
+    Con_Logf(LogSev_Info, "cmd", "displaying %i commands (`help <cmd>` for more info):", ms_cmds.count);
+
+    u32* indices = StrDict_Sort(&ms_cmds, SDictStrCmp, NULL);
+
+    cmddef_t def = { 0 };
+    for (u32 i = 0; i < ms_cmds.count; i++)
+    {
+        u32 index = indices[i];
+        const char* key = ms_cmds.keys[index];
+        Con_Logf(LogSev_Info, "cmd", "\t%s", key);
+    }
+
+    return cmdstat_ok;
+}
+
+static cmdstat_t cmd_echo_fn(i32 argc, const char** argv)
+{
+	char text[1024] = { 0 };
+    for (i32 i = 1; i < argc; ++i)
+    {
+        StrCat(ARGS(text), argv[i]);
+        if ((i + 1) < argc)
+        {
+            StrCat(ARGS(text), " ");
+        }
+    }
+
+    Con_Logf(LogSev_Info, "cmd", text);
+
+    return cmdstat_ok;
+}
```

Comment:
This has the side effect of squashing `echo hello     world`  together, but original Quake does this too. Also, `echo "hello    world" preserves the whitespaces.